src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -128,16 +133,148 @@ static QueryModelVisitor()
+
+				var convert = Expression.Convert(aggregateCall, elementType);
+				aggregateLambda = Expression.Lambda(convert, nullableInputList);
+			}
+
+			_hqlTree.AddPostExecuteTransformer(aggregateLambda);
+		}
+
+		private void AddPostExecuteTransformerForCount()
+		{
+			// Count results have to be summed. No null case to take into account.
+			var elementType = Model.ResultTypeOverride;
+			var inputListType = typeof(IQueryable<>).MakeGenericType(elementType);
+			var inputList = Expression.Parameter(inputListType, "inputList");
+			// Sum has no suitable generic overload, throw in Sum on int, then the code using it
+			// will check and adjust it if it is long instead of int (GetAggregateMethodCall does that).
+			var aggregateCall = GetAggregateMethodCall(ReflectionCache.QueryableMethods.SumOnInt, inputListType, elementType, inputList);
+			_hqlTree.AddPostExecuteTransformer(Expression.Lambda(aggregateCall, inputList));
+		}
+
+		private void AddPostExecuteTransformerForSum()
+		{
+			// Sum is a bit hard, due to an additional mismatch between linq-to-objects and sql semantics on sum, when there are nothing to sum.
+			// linq-to-objets => 0, sql => null
+			// We have to emulate the sql behavior when aggregating polymorphic results.
+			var elementType = Model.ResultTypeOverride;
+			var concreteQueryElementType = elementType;
+			var elementTypeIsNullable = elementType.IsNullable();
+			if (!elementTypeIsNullable)
+			{
+				// Same as in AddPostExecuteTransformerForResultAggregate, override the result type of concrete queries to nullable.
+				// But then, let the nullable sql emulation yield null if there is nothing to sum, and finally try casting the result
+				// non nullable.
+				concreteQueryElementType = typeof(Nullable<>).MakeGenericType(new[] { elementType });
+				_hqlTree.ExecuteResultTypeOverride = concreteQueryElementType;
+			}
+
+			var inputListType = typeof(IQueryable<>).MakeGenericType(concreteQueryElementType);
+			var inputList = Expression.Parameter(inputListType, "inputList");
+			// Sum has no suitable generic overload, throw in Sum on int, then the code using it
+			// will check and adjust it if it is long instead of int (GetAggregateMethodCall does that).
+			var aggregateCall = GetAggregateMethodCall(ReflectionCache.QueryableMethods.SumOnInt, inputListType, concreteQueryElementType, inputList);
+
+			var allMethod = ReflectionCache.EnumerableMethods.AllDefinition.MakeGenericMethod(concreteQueryElementType);
+			var element = Expression.Parameter(concreteQueryElementType, "element");
+			var noSumCondition = Expression.Equal(element, Expression.Constant(null));
+			var allCall = Expression.Call(allMethod, inputList, Expression.Lambda(noSumCondition, element));
+
+			Expression conditionalAggregateCall = Expression.Condition(allCall,
+				Expression.Convert(Expression.Constant(null), concreteQueryElementType),
```

Comment:
Changing it on the above constant too, better to compare both side having the same type.